Intersect.Server/Entities/Player.cs

```diff
@@ -3646,52 +3648,90 @@ public void CraftItem(Guid id)
                     }
                 }
 
-                //Take the items
-                foreach (var c in CraftBase.Get(id).Ingredients)
+                //Return true if the craft was a success
+                if (Randomization.Next(1, 101) > mCraftItem.FailureChance)
                 {
-                    if (!TryTakeItem(c.ItemId, c.Quantity))
+                    //Take the items
+                    foreach (var c in mCraftItem.Ingredients)
+                    {
+                        if (!TryTakeItem(c.ItemId, c.Quantity))
+                        {
+                            for (var i = 0; i < invbackup.Count; i++)
+                            {
+                                Items[i].Set(invbackup[i]);
+                            }
+
+                            PacketSender.SendInventory(this);
+                            CraftId = Guid.Empty;
+
+                            return;
+                        }
+                    }
+
+                    //Give them the craft
+                    var quantity = Math.Max(mCraftItem.Quantity, 1);
+                    var itm = ItemBase.Get(mCraftItem.ItemId);
+                    if (itm == null || !itm.IsStackable)
+                    {
+                        quantity = 1;
```

Comment:
Separate this.

- if not stackable, quantity should be 1 like you have here
- if null, log an error to the console with information about the item id, craft id, and quantity (Log.Error), send some sort of generic error message to the client, and then just `return`

Reply:
I won't know how to do this item, I don't know how to send a log to the server, nor to the client.

Reply:
> I won't know how to do this item, I don't know how to send a log to the server, nor to the client.

here is a little approach/example on how you could handle panda's suggestion:

```
var craftItemId = ItemBase.Get(mCraftItem.ItemId);

if (craftItemId == null)
{
    PacketSender.SendChatMsg(this, Strings.Crafting.craftError, ChatMessageType.Crafting, CustomColors.Alerts.Error);
    Log.Error("[Error]:" + this.Name + " attempted to craft a null item!");
    return;
}
if (!craftItemId.IsStackable)
{
    quantity = 1;
}
```

EDIT: you don't want to hardcode the Log.Error message, this is just an example !

Reply:
@Arufonsu Thanks for your time on this question, but I'm not really going to know how to do it haha, as I said to panda, I didn't do that part of the code, I just replace all craftbase.get() in the method with the variable. If you can send me the code to replace it would help me, because I really don't understand about it. (for example, I saw the code above, I understand, but you said that hardcoded is not ideal, so I have no idea what it would be like. Strings on the server maybe?)

Reply:
```csharp
if (item == null)
{
    PacketSender.SendChatMsg(this, Strings.Errors.UnknownErrorTryAgain, ChatMessageType.Error, CustomColors.Alerts.Errors);
    Log.Error($"Unable to find item descriptor {craftItem.ItemId} for craft {craftItem.Id}.");
    return;
}

if (!craftItemId.IsStackable) {
    quantity = 1;
}
```

Where UnknownErrorTryAgain is `An unknown error occurred, please try again.`